src/crab/ebpf_domain.hpp

```diff
@@ -712,7 +720,7 @@ class ebpf_domain_t final {
     }
 
     NumAbsDomain do_load_stack(NumAbsDomain inv, reg_pack_t target, const linear_expression_t& addr, int width) {
-        if (width == 8) {
+        if (width == 1 || width == 2 || width == 4 || width == 8) {
```

Comment:
If I am not mistaken, this is unsound if the target register is later referred to using a different width (say 8, as in most operations), if the hardware (or actually runtime) does not perform sign extension, and the upper bits are set incorrectly.

Reply:
I didn't quite follow.  Can you elaborate with a potential example scenario about what the case is and what the "unsound" impact is so that I can construct a test case for it?

Reply:
Something like
```
1. r0 = 0
2. store r0 (1 byte)
3. r1 = 1024
4. load r1 (1 byte)
5. access r10[r1 - 4]
```
Does the runtime necessarily perform sign extension or cleans the most significant bytes on partial loads such as on step 4?

If not, the analyzer thinks that after (4), `r1 == 0` but it is not (nor is it bounded by UINT8_MAX), and the access is out of bounds.

Reply:
Yes, it is zero extended.  From the [eBPF VM ISA spec](https://github.com/iovisor/bpf-docs/blob/master/eBPF.md):

0x71 | ldxb dst, [src+off] | dst = *(uint8_t *) (src + off)
-- | -- | --

Note the uint8_t not ssint8_t in particular.  Interpreters implement it that way such as in the [IOVisor uBPF project](https://github.com/iovisor/ubpf/blob/master/vm/ubpf_vm.c#L398):
```
reg[inst.dst] = *(uint8_t *)(uintptr_t)(reg[inst.src] + inst.offset);
```

And similarly a JIT compiler to intel x86 (again from [IOVisor ubpf](https://github.com/iovisor/ubpf/blob/master/vm/ubpf_jit_x86_64.h#L260)) maps to the movzx instruction:
```
        /* movzx */
        emit1(state, 0x0f);
        emit1(state, size == S8 ? 0xb6 : 0xb7);
```

which page 4-140 of the [Intel ISA manual](https://software.intel.com/content/www/us/en/develop/download/intel-64-and-ia-32-architectures-sdm-combined-volumes-1-2a-2b-2c-2d-3a-3b-3c-3d-and-4.html) explains is:

> MOVZX—Move with Zero-Extend
> ...
> Copies the contents of the source operand (register or memory location) to the destination operand (register) and zero extends the value.

So in the example you cite, the eBPF ISA requires `r1 == 0` after (4).

And of course if you try to play the same game with the stack, such as
```
1. r1 = 0
2. *(u64 *)(r10 - 8) = r1 // S.value[-8..-1] = [0]
3. r1 = 1
4. *(u8 *)(r10 - 6) = r1   // S.value[-6..-6] = [1]
5. r2 = *(u64 *)(r10 - 8) // load from S.value[-8..-1]
6. access r10[r2]            // use r2 as index into stack
```
the verifier will correctly fail since step 4 will havoc the S.value[-8..-1] and so step 5 will not think it is 0, but rather that it is unknown.